botocore/docs/bcdoc/docstringparser.py

```diff
@@ -22,30 +21,175 @@ class DocStringParser(html_parser.HTMLParser):
+    meaning that the current_node will be the most recently opened tag. When
+    a tag is closed, the current_node moves up to the parent node.
+    """
+    def __init__(self, doc):
+        self.doc = doc
+        self.head = StemNode()
+        self.current_node = self.head
+        self.unhandled_tags = []
+
+    def add_tag(self, tag, attrs=None, is_start=True):
+        if not self._doc_has_handler(tag, is_start):
+            self.unhandled_tags.append(tag)
+            return
+
+        if is_start:
+            if tag == 'li':
+                node = LineItemNode(attrs)
+            else:
+                node = TagNode(tag, attrs)
+            self.current_node.add_child(node)
+            self.current_node = node
+        else:
+            self.current_node = self.current_node.parent
+
+    def _doc_has_handler(self, tag, is_start):
+        if is_start:
+            handler_name = 'start_%s' % tag
         else:
-            end_space = data[-1].isspace()
-            words = data.split()
-            words = self.doc.translate_words(words)
-            data = ' '.join(words)
+            handler_name = 'end_%s' % tag
+
+        return hasattr(self.doc.style, handler_name)
+
+    def add_data(self, data):
+        self.current_node.add_child(DataNode(data))
+
+    def write(self):
+        self.head.write(self.doc)
+
+
+class Node(object):
+    def __init__(self, parent=None):
+        self.parent = parent
+
+    def write(self, doc):
+        raise NotImplementedError
+
+
+class StemNode(Node):
+    def __init__(self, parent=None):
+        super(StemNode, self).__init__(parent)
+        self.children = []
+
+    def add_child(self, child):
+        child.parent = self
+        self.children.append(child)
+
+    def write(self, doc):
+        self._write_children(doc)
+
+    def _write_children(self, doc):
+        for child in self.children:
+            child.write(doc)
+
+
+class TagNode(StemNode):
+    """
+    A generic Tag node. It will verify that handlers exist before writing.
+    """
+    def __init__(self, tag, attrs=None, parent=None):
+        super(TagNode, self).__init__(parent)
+        self.attrs = attrs
+        self.tag = tag
+
+    def write(self, doc):
+        self._write_start(doc)
+        self._write_children(doc)
+        self._write_end(doc)
+
+    def _write_start(self, doc):
+        handler_name = 'start_%s' % self.tag
+        if hasattr(doc.style, handler_name):
+            getattr(doc.style, handler_name)(self.attrs)
+
+    def _write_end(self, doc):
+        handler_name = 'end_%s' % self.tag
+        if hasattr(doc.style, handler_name):
+            getattr(doc.style, handler_name)()
+
+
+class LineItemNode(TagNode):
+    def __init__(self, attrs=None, parent=None):
+        super(LineItemNode, self).__init__('li', attrs, parent)
+
+    def write(self, doc):
+        self._lstrip()
+        super(LineItemNode, self).write(doc)
+
+    def _lstrip(self):
```

Comment:
I would be fine with removing this method. To simplify the code a bit. It was a little confusing with the two different methods with the one proxying to the other. It would make more sense to have it if _lstrip was not recursive.